ENH: Change SequenceBrowserNode::IndexDisplayDecimals to IndexDisplayFormat

Sunderlandkyl opened the pull request:
re #67

Sunderlandkyl commented:
Still possible to crash if the user inputs format strings like: %s%s%s%s%s.
We could probably use stricter Regex to only match strings with a single %.

Sunderlandkyl commented:
@lassoan Added a step when changing the value to allow the label's fixed width to grow to accommodate the new text, but will never shrink.

Also, the text should already be using a fixed-width font.

Sunderlandkyl commented:
Added a regex to extract the sprintf specifiers from the string:
```
%(?:\d+\$)?[+-]?(?:[ 0]|'.{1})?-?\d*(?:\.\d+)?[fFgGeEs]
```

Only the first match is used to format the string. Remaining prefix and suffix pre/appended to the formatted section.

lassoan commented on SequenceBrowser/MRML/vtkMRMLSequenceBrowserNode.cxx:
```diff
@@ -292,13 +292,12 @@ void vtkMRMLSequenceBrowserNode::ReadXMLAttributes(const char** atts)
       }
       SetIndexDisplayMode(indexDisplayMode);
     }
-    else if (!strcmp(attName, "indexDisplayDecimals"))
```
Could you make this backward compatible? (if number of decimals are specified then generate a format string accordingly)

lassoan commented on SequenceBrowser/Widgets/qMRMLSequenceBrowserSeekWidget.cxx:
```diff
@@ -57,17 +70,90 @@ void qMRMLSequenceBrowserSeekWidgetPrivate::init()
 }
 
+//-----------------------------------------------------------------------------
+QString qMRMLSequenceBrowserSeekWidgetPrivate::formattedIndexValue(int index)
```
If it's not too much work, it would be nice to add a test to check formatting with a couple of non-trivial format strings.

lassoan reviewed (commented):
Looks good, thank you!

Format strings are used in Colors module and various other places in VTK and would be nice to make them safer to use. We could achieve that by adding to vtkAddon a `bool ValidateFormatString(std::string &validatedFormat, const std::string requestedFormat, const std::string& typeString)` method that would convert a format string to one that is guaranteed to be valid for a single specified input. TypeString would be the enforced basic data type (e.g., "s", "f",...) for the input.

lassoan commented:
For now, we can keep `ValidateFormatString` function in this extension. If we want to make Colors module safer or use it elsewhere in Slicer core we could move it there then. It is enough if we have this utility function in as a static member somewhere (probably in the MRML node).

Sunderlandkyl commented:
I've moved the formatting code to vtkMRMLSequenceBrowserNode, and implemented the test/decimals backwards compatibility.

If I implement the ValidateFormatString, do we want to continue pre/appending the non validated parts of the string to the output value?

Ex. With the (ridiculous) index display format of:
```
%%x%t%y%.f5%s
```

Should the expected output still be:
```
%%x%t%y0..00000%s
```

And if it is, should the ValidateFormatString function return the non-formatted prefix/suffix (or the indices of the validatedFormat start/end), or should it require the user to find them using "validatedFormat".

Sunderlandkyl committed:
ENH: Change SequenceBrowserNode::IndexDisplayDecimals to IndexDisplayFormat

- Rather than number of decimals, the user can now specify a sprintf string to format the index value
- Only the first specifier is used (ex. %s%1.2f would only replace %s)
- Available specifiers are: fFgGeEs

- When value is displayed in label, the fixed width is calculated as the max of the width of the current value and the current width of the label
- Allows the label to grow in size, but never shrink
- Label size resets when user switches SequenceBrowserNodes

re #67

lassoan commented:
Good point. Maybe we should return prefix and suffix (unmatched parts of the regexp) as additional outputs of ValidateFormatString.

Sunderlandkyl commented:
ValidateFormatString added.

lassoan commented:
Thank you!
